cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java

```diff
@@ -0,0 +1,280 @@
+          Collection<QueueBarrier> groupBarriers = barriers.get(groupId);
+          PeekingIterator<QueueBarrier> barrierIterator = Iterators.peekingIterator(groupBarriers.iterator());
+          while (barrierIterator.hasNext()) {
+            QueueBarrier start = barrierIterator.next();
+            QueueBarrier end = barrierIterator.hasNext() ? barrierIterator.peek() : null;
+
+            System.out.printf("Scanning section %d/%d...", currentSection, barriers.size()).println();
+            scanQueue(stateStore.getTransaction(), queueName, start, end, stats);
+            System.out.printf("Current results: %s", stats.getReport()).println();
+            currentSection++;
+          }
+          System.out.println("Scanning complete");
+        }
+
+        System.out.printf("Total results: %s", stats.getReport()).println();
+      }
+    }, stateStore);
+  }
+
+  private void scanQueue(Transaction tx, QueueName queueName, QueueBarrier start,
+                         @Nullable QueueBarrier end, QueueStatistics outStats) throws IOException {
+
+    byte[] queueRowPrefix = QueueEntryRow.getQueueRowPrefix(queueName);
+
+    ConsumerGroupConfig consumerConfig = start.getGroupConfig();
+    System.out.printf("Got consumer group config: %s", consumerConfig).println();
+
+    if (!queueClientFactory.tableExists(queueName)) {
+      System.out.printf("Found no flow queue table for %s", queueName).println();
+      return;
+    }
+
+    HBaseQueueAdmin admin = queueClientFactory.ensureTableExists(queueName);
+    TableId tableId = admin.getDataTableId(queueName, QueueConstants.QueueType.SHARDED_QUEUE);
+    HTable hTable = queueClientFactory.createHTable(tableId);
+
+    System.out.printf("Looking at HBase table: %s", Bytes.toString(hTable.getTableName())).println();
+
+    byte[] stateColumnName = Bytes.add(QueueEntryRow.STATE_COLUMN_PREFIX,
+                                       Bytes.toBytes(consumerConfig.getGroupId()));
+
+
+    int distributorBuckets = queueClientFactory.getDistributorBuckets(hTable.getTableDescriptor());
+    ShardedHBaseQueueStrategy queueStrategy = new ShardedHBaseQueueStrategy(distributorBuckets);
+
+    Scan scan = new Scan();
+    scan.setStartRow(start.getStartRow());
+    if (end != null) {
+      scan.setStopRow(end.getStartRow());
+    }
+    scan.addColumn(QueueEntryRow.COLUMN_FAMILY, QueueEntryRow.DATA_COLUMN);
+    scan.addColumn(QueueEntryRow.COLUMN_FAMILY, QueueEntryRow.META_COLUMN);
+    scan.addColumn(QueueEntryRow.COLUMN_FAMILY, stateColumnName);
+    scan.setMaxVersions(1);
+
+    System.out.printf("Scanning section with scan: %s", scan.toString()).println();
+    for (int instanceId = 0; instanceId < consumerConfig.getGroupSize(); instanceId++) {
+      System.out.printf("Processing instance %d", instanceId).println();
+      ConsumerConfig consConfig = new ConsumerConfig(consumerConfig, instanceId);
+      QueueScanner scanner = queueStrategy.createScannerWithActualRowKeys(consConfig, hTable, scan, 100);
```

Comment:
So the `Scan` contains barrier start/stop row, which is never be a actual row key, but rather just the un-salted, un-partitioned rowkey (Hash + QueueName + Write pointer + counter). How does it work now?